include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h

```diff
@@ -1182,37 +1205,35 @@ __parallel_find_or(oneapi::dpl::__internal::__device_backend_tag, _ExecutionPoli
                 [=](sycl::nd_item</*dim=*/1> __item_id) {
                     auto __local_idx = __item_id.get_local_id(0);
 
-                    __dpl_sycl::__atomic_ref<_AtomicType, sycl::access::address_space::global_space> __found(
-                        *__dpl_sycl::__get_accessor_ptr(__temp_acc));
-                    __dpl_sycl::__atomic_ref<_AtomicType, sycl::access::address_space::local_space> __found_local(
-                        *__dpl_sycl::__get_accessor_ptr(__temp_local));
-
-                    // 1. Set initial value to local atomic
-                    if (__local_idx == 0)
-                        __found_local.store(__init_value);
-                    __dpl_sycl::__group_barrier(__item_id);
+                    // 1. Set initial value to local found state
+                    _AtomicType __found_local = __init_value;
 
-                    // 2. Find any element that satisfies pred and set local atomic value to global atomic
+                    // 2. Find any element that satisfies pred
                     __pred(__item_id, __n_iter, __wgroup_size, __found_local, __brick_tag, __rngs...);
-                    __dpl_sycl::__group_barrier(__item_id);
 
-                    // Set local atomic value to global atomic
-                    if (__local_idx == 0)
+                    // 3. Reduce over group: find __dpl_sycl::__minimum (for the __parallel_find_forward_tag),
+                    // find __dpl_sycl::__maximum (for the __parallel_find_backward_tag)
+                    // or calculate the sum with __dpl_sycl::__plus (for the __parallel_or_tag)
+                    // inside all our group items
+                    __found_local = __dpl_sycl::__reduce_over_group(__item_id.get_group(), __found_local,
+                                                                    typename _BrickTag::FoundLocalReduceOp{});
```

Comment:
As mentioned offline, we may find better performance by using manually written subgroup reductions followed by a single subgroup reducing the subgroups to a single value, but this can be pursued in a later PR if we decide to do it. I'm not sure how significant it would be.

Or we can look to improve `__reduce_over_group` itself.